### PR TITLE
DDF-2413 Refactored CSW to use new taxonomy and cleaned up code

### PR DESCRIPTION
**Pull request by bdeining**
#### What does this PR do?
This PR refactors the entire CSW project.  It:
- aligns everything to the new taxonomy
- removes dead and unused code
- java 8ifies code

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@glenhein @jrnorth 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@jlcsmith
@kcwire
#### How should this be tested?
Build / Install DDF.  Configure a CSW Federated source and observe correct behavior.  Perform CSW operations (create / update / delete) and query on the CSW Endpoint and observer proper behavior.
#### What are the relevant tickets?
https://codice.atlassian.net/browse/DDF-2413
#### Checklist:
- [ ] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswConstants.java [hunk 1]
```diff
@@ -22,266 +22,699 @@
...
+    public static final String OWS_LOWER_CORNER = "LowerCorner";
 
-    String OWS_UPPER_CORNER = "UpperCorner";
+    public static final double DEGREES_TO_RADIANS = Math.PI / 180.0;
 
-    String OWS_LOWER_CORNER = "LowerCorner";
+    public static final double RADIANS_TO_DEGREES = 1 / DEGREES_TO_RADIANS;
 
-    double DEGREES_TO_RADIANS = Math.PI / 180.0;
+    public static final double EARTH_MEAN_RADIUS_METERS = 6371008.7714;
```
Comment (glenhein) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswConstants.java]: I wondering if you should drop in a javadoc that indicates the official source of this value.

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswMarshallHelper.java [hunk 2]
```diff
@@ -86,31 +86,37 @@ static void writeTemporalData(HierarchicalStreamWriter writer, MarshallingContex
             MetacardImpl metacard) {
         StringBuilder sb = new StringBuilder();
         sb.append(ISODateTimeFormat.dateTime()
-                .print(((Date) metacard.getEffectiveDate()).getTime()))
+                .print((metacard.getEffectiveDate()).getTime()))
```
Comment (glenhein) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswMarshallHelper.java]: extra parens?

**Comment by glenhein** [comment]
👍

**Comment by bdeining** [comment]
This PR depends on https://github.com/codice/ddf/pull/1219

**Comment by bdeining** [comment]
This PR depends on https://github.com/codice/ddf/pull/1219

**Comment by bdeining** [comment]
This PR depends on https://github.com/codice/ddf/pull/1219

**Comment by bdeining** [comment]
This PR depends on https://github.com/codice/ddf/pull/1219

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/converter/DefaultCswRecordMap.java [hunk 3]
```diff
@@ -51,108 +54,154 @@
...
+        localNameMap.put(CswConstants.CSW_PUBLISHER, Contact.PUBLISHER_NAME);
+        localNameMap.put(CswConstants.CSW_CONTRIBUTOR, Contact.CONTRIBUTOR_NAME);
+        localNameMap.put(CswConstants.CSW_CREATOR, Contact.CREATOR_NAME);
+        localNameMap.put(CswConstants.CSW_RELATION, Core.RESOURCE_DOWNLOAD_URL);
+        localNameMap.put(CswConstants.CSW_TABLE_OF_CONTENTS, Core.DESCRIPTION);
+        localNameMap.put(CswConstants.CSW_ABSTRACT, Core.DESCRIPTION);
+        localNameMap.put(CswConstants.CSW_DESCRIPTION, Core.DESCRIPTION);
+        localNameMap.put(CswConstants.CSW_SUBJECT, Topic.CATEGORY);
+        localNameMap.put(CswConstants.OWS_BOUNDING_BOX, Core.LOCATION);
+        localNameMap.put(CswConstants.CSW_SOURCE, Core.RESOURCE_URI);
```
Comment (jlcsmith) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/converter/DefaultCswRecordMap.java]: CSW_SOURCE was added already.  This line can be removed.

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/converter/DefaultCswRecordMap.java [hunk 4]
```diff
@@ -51,108 +54,154 @@
...
+        qNameMap.put(CswConstants.CSW_PUBLISHER_QNAME, Contact.PUBLISHER_NAME);
+        qNameMap.put(CswConstants.CSW_CONTRIBUTOR_QNAME, Contact.CONTRIBUTOR_NAME);
+        qNameMap.put(CswConstants.CSW_CREATOR_QNAME, Contact.CREATOR_NAME);
+        qNameMap.put(CswConstants.CSW_RELATION_QNAME, Core.RESOURCE_DOWNLOAD_URL);
+        qNameMap.put(CswConstants.CSW_ABSTRACT_QNAME, Core.DESCRIPTION);
+        qNameMap.put(CswConstants.CSW_TABLE_OF_CONTENTS_QNAME, Core.DESCRIPTION);
+        qNameMap.put(CswConstants.CSW_DESCRIPTION_QNAME, Core.DESCRIPTION);
+        qNameMap.put(CswConstants.CSW_SUBJECT_QNAME, Topic.CATEGORY);
+        qNameMap.put(CswConstants.OWS_BOUNDING_BOX_QNAME, Core.LOCATION);
+        qNameMap.put(CswConstants.CSW_SOURCE_QNAME, Core.RESOURCE_URI);
```
Comment (jlcsmith) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/converter/DefaultCswRecordMap.java]: same thing here

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswMarshallHelper.java [hunk 5]
```diff
@@ -86,31 +86,35 @@ static void writeTemporalData(HierarchicalStreamWriter writer, MarshallingContex
...
                 if (ad == null) {
-                    ad = new AttributeDescriptorImpl(attrName,
-                            false,
-                            false,
-                            false,
-                            false,
-                            BasicTypes.STRING_TYPE);
+                    /*  Backwards Compatibility */
+                    ad = BasicTypes.BASIC_METACARD.getAttributeDescriptor(attrName);
+                    if (ad == null) {
```
Comment (jlcsmith) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswMarshallHelper.java]: Why is this part needed (creation of new attribute descriptors)?

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java [hunk 6]
```diff
@@ -132,179 +128,109 @@ public static void removeExistingAttributes(Map<String, String> cswAttrMap,
...
-                    attributes.put(name, new AttributeImpl(name, value));
-                }
-
-                if (BasicTypes.GEO_TYPE.getAttributeFormat()
-                        .equals(attributeDescriptor.getType()
-                                .getAttributeFormat())) {
-                    mc.setLocation((String) value);
                 }
+            /* Set Content Type for backwards compatibility */
+            } else if (name.equals("type")) {
```
Comment (jlcsmith) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: shouldn't we be using a constant here for "type"?

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/converter/DefaultCswRecordMap.java (hunk 3 above)
Comment (bdeining) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/converter/DefaultCswRecordMap.java]: Removed

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/converter/DefaultCswRecordMap.java (hunk 4 above)
Comment (bdeining) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/converter/DefaultCswRecordMap.java]: Removed

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswMarshallHelper.java (hunk 5 above)
Comment (bdeining) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswMarshallHelper.java]: This isn't needed and was removed

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswRecordConverter.java [hunk 7]
```diff
@@ -306,19 +309,22 @@ public static Serializable convertStringValueToMetacardValue(
      * Converts the CSW record property {@code reader} is currently at to the specified Metacard
      * attribute format.
      *
-     * @param attributeFormat the {@link AttributeType.AttributeFormat} corresponding to the type that the value
-     *                        in {@code reader} should be converted to
-     * @param reader          the reader at the element whose value you want to convert
-     * @param cswAxisOrder    the order of the coordinates in the XML being read by {@code reader}
+     * @param attributeName the {@link AttributeType.AttributeFormat} corresponding to the type that the value
```
Comment (bdeining) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswRecordConverter.java]: fixed

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java (hunk 6 above)
Comment (bdeining) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: changed to constant

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswConstants.java [hunk 8]
```diff
@@ -22,266 +22,699 @@
...
+                CSW_SOURCE_QNAME,
+                OWS_BOUNDING_BOX_QNAME);
+    }
+
+    private static QName createDublinCoreQName(final String field) {
+        return new QName(DUBLIN_CORE_SCHEMA, field, DUBLIN_CORE_NAMESPACE_PREFIX);
+    }
+
+    private static QName createDublinCoreTermQName(final String field) {
+        return new QName(CswConstants.DUBLIN_CORE_TERMS_SCHEMA,
```
Comment (jrnorth) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswConstants.java]: Don't need the `CswConstants.`

File: catalog/spatial/csw/spatial-csw-endpoint/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswQueryFactoryTest.java [hunk 9]
```diff
@@ -266,7 +271,7 @@ public void testPostGetRecordsDistributedSearchSpecificSources()
         grr.setDistributedSearch(distributedSearch);
 
         QueryType query = new QueryType();
-        List<QName> typeNames = new ArrayList<QName>();
+        List<QName> typeNames = new ArrayList();
```
Comment (jrnorth) [inline catalog/spatial/csw/spatial-csw-endpoint/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswQueryFactoryTest.java]: Should keep the `<>`.

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java [hunk 10]
```diff
@@ -132,179 +128,109 @@ public static void removeExistingAttributes(Map<String, String> cswAttrMap,
...
-                        .getAttributeFormat(), reader, cswAxisOrder);
-            }
 
-            if (null != value) {
-                if (attributeDescriptor.isMultiValued()) {
-                    if (attributes.containsKey(name)) {
-                        AttributeImpl attribute = (AttributeImpl) attributes.get(name);
-                        attribute.addValue(value);
+            if (DefaultCswRecordMap.hasDefaultMetacardFieldFor(name)) {
+                String test = DefaultCswRecordMap.getDefaultMetacardFieldFor(name);
```
Comment (jrnorth) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: Should probably name this something more descriptive than `test`.

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java [hunk 11]
```diff
@@ -132,179 +128,109 @@ public static void removeExistingAttributes(Map<String, String> cswAttrMap,
...
+                    String attributeName = attributeDescriptor.getName();
+
+                    if (attributeName.equals(Core.THUMBNAIL)) {
+                        String uri = new String((byte[]) value, StandardCharsets.UTF_8);
+                        URL url;
+                        InputStream is = null;
+                        try {
+                            url = new URL(uri);
+                            is = url.openStream();
+                            mc.setThumbnail(IOUtils.toByteArray(url.openStream()));
```
Comment (jrnorth) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: `IOUtils.toByteArray(is)`?

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java [hunk 12]
```diff
@@ -132,179 +128,109 @@ public static void removeExistingAttributes(Map<String, String> cswAttrMap,
...
+                        } catch (IOException e) {
+                            LOGGER.debug(
```
Comment (jrnorth) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: `LOGGER.debug("Error setting thumbnail data on metacard: {}", value, e);`

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java [hunk 13]
```diff
@@ -132,179 +128,109 @@ public static void removeExistingAttributes(Map<String, String> cswAttrMap,
...
+
+                    String attributeName = attributeDescriptor.getName();
+
+                    if (attributeName.equals(Core.THUMBNAIL)) {
+                        String uri = new String((byte[]) value, StandardCharsets.UTF_8);
+                        URL url;
+                        InputStream is = null;
+                        try {
+                            url = new URL(uri);
+                            is = url.openStream();
```
Comment (stustison) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: Why are we opening a URL connection to get the thumbnail? If we really need to do this, we should at least be verifying that what we get back is actually an image and that it isn't huge. Just dumping url.openStream() to a byte array and saving it is almost certainly going to cause a heap space issue.

Comment (stustison) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: To be clear, this is 👎  until the input is verified.

Comment (jlcsmith) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: @stustison - the thumbnail isn't included in the Csw response and that is why we open a connection to go get it.  You definitely make a valid point about checking the response.  As a note, this wasn't introduced in this PR.  In fact, the CSW code has been doing this for a long time.  I'd suggest we open a separate ticket for this issue.

Comment (stustison) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: OK that is fine. 👍  on condition of a ticket being written.

Comment (bdeining) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: https://codice.atlassian.net/browse/DDF-2476

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswConstants.java (hunk 8 above)
Comment (bdeining) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswConstants.java]: Removed all lingering occurrences of `CswConstants`

File: catalog/spatial/csw/spatial-csw-endpoint/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswQueryFactoryTest.java (hunk 9 above)
Comment (bdeining) [inline catalog/spatial/csw/spatial-csw-endpoint/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswQueryFactoryTest.java]: Fixed

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java (hunk 10 above)
Comment (bdeining) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: Renamed

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java (hunk 11 above)
Comment (bdeining) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: Done

File: catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java (hunk 12 above)
Comment (bdeining) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/converter/CswUnmarshallHelper.java]: Fixed

**Comment by jlcsmith** [comment]
branch needs to be rebased please

**Comment by bdeining** [comment]
@jlcsmith rebased
